### ISSUE TITLE
Porting to 6.7.17+?

### ISSUE THREAD

**Issue by praiskup** [plv8/plv8#281]
In Fedora 28, there's now v8-devel-6.7.17-4.fc28.x86_64;  and the plv8 module fails to build:

```
plv8.cc:193:11: error: ‘Protection’ has not been declared
           Protection protection) override {}
           ^~~~~~~~~~
plv8.cc:149:8: error: ‘void* Plv8ArrayBufferAllocator::Reserve(size_t)’ marked ‘override’, but does not override
  void* Reserve(size_t length) override {
        ^~~~~~~
plv8.cc:188:6: error: ‘void Plv8ArrayBufferAllocator::Free(void*, size_t, v8::ArrayBuffer::Allocator::AllocationMode)’ marked ‘override’, but does not override
 void Free(void *data, size_t size, v8::ArrayBuffer::Allocator::AllocationMode) override {
      ^~~~
plv8.cc:192:7: error: ‘void Plv8ArrayBufferAllocator::SetProtection(void*, size_t, int)’ marked ‘override’, but does not override
  void SetProtection(void* data, size_t length,
       ^~~~~~~~~~~~~
plv8.cc: In function ‘void _PG_init()’:
plv8.cc:348:7: error: ‘InitializeICU’ is not a member of ‘v8::V8’
   V8::InitializeICU();
       ^~~~~~~~~~~~~
plv8.cc:351:7: error: ‘InitializeICU’ is not a member of ‘v8::V8’
   V8::InitializeICU(plv8_icu_data);
       ^~~~~~~~~~~~~
plv8.cc: In function ‘void signal_handler(int)’:
plv8.cc:552:10: error: ‘TerminateExecution’ is not a member of ‘v8::V8’
  v8::V8::TerminateExecution(plv8_isolate);
          ^~~~~~~~~~~~~~~~~~
plv8.cc: In function ‘v8::Local<v8::Value> DoCall(v8::Handle<v8::Function>, v8::Handle<v8::Object>, int, v8::Handle<v8::Value>*)’:
plv8.cc:564:12: error: no matching function for call to ‘v8::TryCatch::TryCatch()’
  TryCatch  try_catch;
            ^~~~~~~~~
In file included from plv8.h:12,
                 from plv8.cc:8:
/usr/include/v8.h:8301:12: note: candidate: ‘v8::TryCatch::TryCatch(v8::Isolate*)’
   explicit TryCatch(Isolate* isolate);
            ^~~~~~~~
/usr/include/v8.h:8301:12: note:   candidate expects 1 argument, 0 provided
plv8.cc: In function ‘Datum CallTrigger(FunctionCallInfo, plv8_exec_env*)’:
plv8.cc:882:13: error: no matching function for call to ‘v8::TryCatch::TryCatch()’
  TryCatch   try_catch;
             ^~~~~~~~~
In file included from plv8.h:12,
                 from plv8.cc:8:
/usr/include/v8.h:8301:12: note: candidate: ‘v8::TryCatch::TryCatch(v8::Isolate*)’
   explicit TryCatch(Isolate* isolate);
            ^~~~~~~~
/usr/include/v8.h:8301:12: note:   candidate expects 1 argument, 0 provided
plv8.cc: In function ‘char* CompileDialect(const char*, Dialect)’:
plv8.cc:1225:12: error: no matching function for call to ‘v8::TryCatch::TryCatch()’
  TryCatch  try_catch;
            ^~~~~~~~~
In file included from plv8.h:12,
                 from plv8.cc:8:
/usr/include/v8.h:8301:12: note: candidate: ‘v8::TryCatch::TryCatch(v8::Isolate*)’
   explicit TryCatch(Isolate* isolate);
            ^~~~~~~~
/usr/include/v8.h:8301:12: note:   candidate expects 1 argument, 0 provided
plv8.cc: In function ‘v8::Local<v8::Function> CompileFunction(v8::Persistent<v8::Context>&, const char*, int, const char**, const char*, bool, bool, Dialect)’:
plv8.cc:1399:12: error: no matching function for call to ‘v8::TryCatch::TryCatch()’
  TryCatch  try_catch;
            ^~~~~~~~~
In file included from plv8.h:12,
                 from plv8.cc:8:
/usr/include/v8.h:8301:12: note: candidate: ‘v8::TryCatch::TryCatch(v8::Isolate*)’
   explicit TryCatch(Isolate* isolate);
            ^~~~~~~~
/usr/include/v8.h:8301:12: note:   candidate expects 1 argument, 0 provided
plv8.cc: In function ‘v8::Local<v8::Function> find_js_function(Oid)’:
plv8.cc:1458:14: error: no matching function for call to ‘v8::TryCatch::TryCatch()’
   TryCatch   try_catch;
              ^~~~~~~~~
In file included from plv8.h:12,
                 from plv8.cc:8:
/usr/include/v8.h:8301:12: note: candidate: ‘v8::TryCatch::TryCatch(v8::Isolate*)’
   explicit TryCatch(Isolate* isolate);
            ^~~~~~~~
/usr/include/v8.h:8301:12: note:   candidate expects 1 argument, 0 provided
plv8.cc: In function ‘void GetGlobalContext(v8::Persistent<v8::Context>&)’:
plv8.cc:1587:15: error: no matching function for call to ‘v8::TryCatch::TryCatch()’
    TryCatch   try_catch;
               ^~~~~~~~~
In file included from plv8.h:12,
                 from plv8.cc:8:
/usr/include/v8.h:8301:12: note: candidate: ‘v8::TryCatch::TryCatch(v8::Isolate*)’
   explicit TryCatch(Isolate* isolate);
            ^~~~~~~~
/usr/include/v8.h:8301:12: note:   candidate expects 1 argument, 0 provided
plv8.cc: In function ‘v8::Local<v8::ObjectTemplate> GetGlobalObjectTemplate()’:
plv8.cc:1666:53: error: no matching function for call to ‘v8::ObjectTemplate::New()’
   Local<ObjectTemplate> templ = ObjectTemplate::New();
                                                     ^
In file included from plv8.h:12,
                 from plv8.cc:8:
/usr/include/v8.h:5772:32: note: candidate: ‘static v8::Local<v8::ObjectTemplate> v8::ObjectTemplate::New(v8::Isolate*, v8::Local<v8::FunctionTemplate>)’
   static Local<ObjectTemplate> New(
                                ^~~
/usr/include/v8.h:5772:32: note:   candidate expects 2 arguments, 0 provided
/usr/include/v8.h:5979:32: note: candidate: ‘static v8::Local<v8::ObjectTemplate> v8::ObjectTemplate::New(v8::internal::Isolate*, v8::Local<v8::FunctionTemplate>)’
   static Local<ObjectTemplate> New(internal::Isolate* isolate,
                                ^~~
/usr/include/v8.h:5979:32: note:   candidate expects 2 arguments, 0 provided
plv8.cc:1681:53: error: no matching function for call to ‘v8::ObjectTemplate::New()’
   Handle<ObjectTemplate> plv8 = ObjectTemplate::New();
                                                     ^
In file included from plv8.h:12,
                 from plv8.cc:8:
/usr/include/v8.h:5772:32: note: candidate: ‘static v8::Local<v8::ObjectTemplate> v8::ObjectTemplate::New(v8::Isolate*, v8::Local<v8::FunctionTemplate>)’
   static Local<ObjectTemplate> New(
                                ^~~
/usr/include/v8.h:5772:32: note:   candidate expects 2 arguments, 0 provided
/usr/include/v8.h:5979:32: note: candidate: ‘static v8::Local<v8::ObjectTemplate> v8::ObjectTemplate::New(v8::internal::Isolate*, v8::Local<v8::FunctionTemplate>)’
   static Local<ObjectTemplate> New(internal::Isolate* isolate,
                                ^~~
/usr/include/v8.h:5979:32: note:   candidate expects 2 arguments, 0 provided
plv8.cc: In member function ‘Datum Converter::ToDatum(v8::Handle<v8::Value>, Tuplestorestate*)’:
plv8.cc:1813:12: error: no matching function for call to ‘v8::TryCatch::TryCatch()’
  TryCatch  try_catch;
            ^~~~~~~~~
In file included from plv8.h:12,
                 from plv8.cc:8:
/usr/include/v8.h:8301:12: note: candidate: ‘v8::TryCatch::TryCatch(v8::Isolate*)’
   explicit TryCatch(Isolate* isolate);
            ^~~~~~~~
/usr/include/v8.h:8301:12: note:   candidate expects 1 argument, 0 provided
make: *** [Makefile:91: plv8.o] Error 1
```

**Comment by JerrySievert** [comment]
correct.  v8's API changes quite often, and thus software that embeds it needs to be tied to specific ranges of versions of v8 in order to work.

plv8 is tied to 6.4.388.44, which was the same version that the LTS node.js 8 was tied to.  when node releases another LTS, plv8 will as well.

as an aside, you should not be using a shared version of v8, the v8 team has taken steps to make it difficult to support, and pretty much every embedded compile of v8 is used differently by the host software - making it extremely unreliable to use in that manner.  for instance, building with or without snapshots changes how v8 performs, as well as external startups.

**Comment by praiskup** [comment]
> as an aside, you should not be using a shared version of v8, the v8 team has taken steps to make it difficult to support

Ah, OK, thanks.  This is probably show-stopper for plv8 in Fedora then :-( since I'm not able to take care of the whole bundled V8 stuff.

**Comment by JerrySievert** [comment]
it seems that every distribution supports wildly different versions of v8 that are incompatible with each other.  how does fedora handle distribution of nodejs?

**Comment by praiskup** [comment]
Fedora seems to update V8 from time to time, now it was from `6.2` to `6.7`;  I don't know more details.  Here are the relevant Fedora bugs:
https://bugzilla.redhat.com/1584453
https://bugzilla.redhat.com/1585159
https://bugzilla.redhat.com/1585212

**Comment by JerrySievert** [comment]
6.4.388.14 was the version that fixed the spectre/meltdown issues, fyi.  any sane distribution should be at least on 6.4.388.14.

still curious how fedora handles nodejs, since fedora 28 appears to be supporting 8.11, which is tied to 6.4.388.

**Comment by JerrySievert** [comment]
ah, I see, fedora 28 is shipping with 6.2.

**Comment by praiskup** [comment]
It used to, but now F28 got update to `6.7.17`

**Comment by JerrySievert** [comment]
although the rpm shows it as bundled.  do you know which it is?

**Comment by praiskup** [comment]
What RPM do you have in hand?  This is mine:
```
$ rpm -qi v8
Name        : v8
Epoch       : 1
Version     : 6.7.17
Release     : 4.fc28
Architecture: x86_64
Install Date: Fri 01 Jun 2018 05:56:28 PM CEST
Group       : System Environment/Libraries
Size        : 40192190
License     : BSD
Signature   : RSA/SHA256, Tue 15 May 2018 08:15:49 PM CEST, Key ID e08e7e629db62fb1
Source RPM  : v8-6.7.17-4.fc28.src.rpm
Build Date  : Tue 15 May 2018 04:05:29 PM CEST
Build Host  : buildvm-31.phx2.fedoraproject.org
Relocations : (not relocatable)
Packager    : Fedora Project
Vendor      : Fedora Project
URL         : https://chromium.googlesource.com/v8/v8/
Bug URL     : https://bugz.fedoraproject.org/v8
Summary     : JavaScript Engine
Description :
V8 is Google's open source JavaScript engine. V8 is written in C++ and is used
in Google Chrome, the open source browser from Google. V8 implements ECMAScript
```

**Comment by JerrySievert** [comment]
I was looking at the nodejs package itself.

what you posted still doesn't show the version of v8 that nodejs is using.  looking at the rpm file itself, it's saying that it is bundled, which is what I expected.

you can verify this by installing nodejs, and looking at `process.versions`:

```
{ http_parser: '2.7.0',
  node: '8.9.4',
  v8: '6.1.534.50',
  uv: '1.15.0',
  zlib: '1.2.11',
  ares: '1.10.1-DEV',
  modules: '57',
  nghttp2: '1.25.0',
  openssl: '1.0.2n',
  icu: '59.1',
  unicode: '9.0',
  cldr: '31.0.1',
  tz: '2017b' }
```

(my copy of node is out of date locally, as you see).

If it is bundled, then your v8 maintainer is only maintaining v8 for chrome, and not nodejs, which means that bundling v8 is not unprecedented.

**Comment by JerrySievert** [comment]
According to the v8 team, v8 should only ever be bundled, and never distributed by itself - every package uses its own configuration and those configurations are not compatible with each other. 

Plv8 should not be using a shared package and is designed to be bundling it as well. 

The v8 team has taken great steps to make sure that it does not get compiled as a shared library by changing the build system, and every release that we would consider minor (6.3 to 6.4, etc) can be a breaking change. 

Long story short: don’t try to build a shared version, it’s a disservice to your users.  Bundle instead.

**Comment by praiskup** [comment]
Meh, what a modern approach :-( thanks for the info Jerry.

**Comment by JerrySievert** [comment]
Every compile flag changes the behavior significantly, and each application embedding v8 has its own requirements, so it doesn’t make sense to try to require a single shared library. Likely this is one of the major reasons why distro maintainers have tried to stick with 3.14, even though it is ancient. 

As an aside, v8 also has its own compiler suite. 

Just bundle, and your maintanence life will be much easier.

**Comment by JerrySievert** [comment]
As for the ticket itself, plv8 will support a 6.7 or higher version at some point in the future when it makes sense, so going to go ahead and close this ticket.

**Comment by praiskup** [comment]
> Every compile flag changes the behavior significantly

.. as everywhere (not only v8).   And still there are mature languages where bundling is not needed.

> Just bundle, and your maintanence life will be much easier.

Definitely not.  It's simply too soon for me to take the responsibility for plv8 in Fedora.

Jerry, don't you have link to the "don't bundle" upstream recommendation?

**Comment by JerrySievert** [comment]
https://github.com/plv8/plv8/issues/251

Has links to official communication from the v8 team. It’s also a frequent topic on the v8 mailing list. @christophberg is the Debian maintainer, btw, and is able to speak with authority for how Debian has been handling it.

**Comment by ChristophBerg** [comment]
@praiskup: Fwiw, the situation for Debian is exactly the same, except we are still stuck with 3.14.5... It's a sad story :(

**Comment by praiskup** [comment]
@ChristophBerg I can see we still have 314 in Fedora, too.  What version of plv8 are you shipping, and downstream patches are you using?  I admit that I'd go the same way... The git link (http://anonscm.debian.org/gitweb/?p=pkg-postgresql/plv8.git) on
https://packages.debian.org/source/jessie/plv8 doesn't work...

**Comment by ChristophBerg** [comment]
The exact version is 3.14.5.8 and has not been changed for a really long time (same in wheezy/jessie/stretch aka Debian 6/7/8). There's been a note in our release notes that the JS universe is complicated and that because of that, v8 and all of its reverse-dependencies are not security-supported (https://www.debian.org/releases/stable/amd64/release-notes/ch-information.en.html#libv8).
In the meantime, the package has accumulated 4 RC bugs (https://bugs.debian.org/cgi-bin/pkgreport.cgi?repeatmerged=no&src=libv8-3.14), so it won't be part of the next release.
I've effectively stopped plv8 maintenance because of that - I guess the best action would really be to admit defeat and have the package removed from Debian unstable as well.

(Re the git link, Debian has moved to a Gitlab instance: https://salsa.debian.org/postgresql/plv8)

**Comment by praiskup** [comment]
Thanks for the info, so FTR, I proposed removal from Fedora:
https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/5ASGLU7T5CQ4JP63SHYSBD7VOLERU2AP/
And coincidentally, 5 minutes earlier the whole v8 removal was announced:
https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/DI4Q5526MVI5KS7OG4PH37QFK6KCDAY2/
